src/index/MetaDataHandler.h

```diff
@@ -0,0 +1,199 @@
+    std::pair<Id, const FullRelationMetaData&> operator*() const {
+      // make sure that we do not conflict with the empty key
+      AD_CHECK(_id != size_t(-1));
+      return std::make_pair(_id, std::cref(*_it));
+    }
+
+    // _____________________________________________________________
+    Iterator(Id id, const typename ExtVec::const_iterator& it, const ExtVec* const vec): _id(id), _it(it), _vec(vec){}
+
+
+    // ________________________________________________
+    Iterator& operator++() {
+      ++_id;
+      ++_it;
+      goToNexValidEntry();
+
+      return *this;
+    }
+
+    // ___________________________________________________
+    void goToNexValidEntry() {
+      while(_it != _vec->cend() && (*_it) == emptyMetaData) {
+	++_id;
+	++_it;
+      }
+    }
+
+    // ________________________________________________
+    Iterator operator++(int) {
+      Iterator old(*this);
+      ++_id;
+      ++_it;
+      goToNexValidEntry();
+      return old;
+    }
+
+    // _______________________________________________
+    bool operator==(const Iterator& other) {
+      return _it == other._it;
+    }
+
+    // _______________________________________________
+    bool operator!=(const Iterator& other) {
+      return _it != other._it;
+    }
+
+   private:
+
+    
+    Id _id;
+    typename ExtVec::const_iterator _it;
+    const ExtVec* const _vec;
+
+    const FullRelationMetaData emptyMetaData = FullRelationMetaData::empty;
+  };
+
+}
+
+// _____________________________________________________________________
+template <class M>
+class MetaDataWrapperDense {
+ public:
+  using Iterator = ExtVecWrapperImpl::Iterator<M>;
+
+  // _________________________________________________________
+  MetaDataWrapperDense() = default;
+
+  // Templated setup version
+  // Arguments are passsed through to template argument M.
+  // TODO<joka921>: enable_if  for better error messages
+  template<typename... Args>
+  void setup(Args... args) {
+    _size = 0;
+    _vec = M(args...);
+    //for (const auto& el : _vec) {
+    for (auto it = _vec.begin(); it != _vec.end(); ++it) {
+      const auto& el = *it;
+      _size += (el != emptyMetaData);
+    }
+  }
+  
+  // ___________________________________________________________
+  size_t size() const {return _size;}
+
+  // __________________________________________________________________
+  const Iterator cbegin() const {
+    Iterator it(0, _vec.begin(), &_vec);
+    it.goToNexValidEntry();
+    return it;
+  }
+
+  // __________________________________________________________________________
+  const Iterator cend() const {
+    return Iterator(_vec.size(), _vec.end(), &_vec);
+  }
+
+  // ____________________________________________________________
+  void set(Id id, const FullRelationMetaData& value) {
+    AD_CHECK(id < _vec.size());
+    bool previouslyEmpty = _vec[id] == emptyMetaData;
+
+    // check that we never insert the empty key
+    AD_CHECK(value != emptyMetaData);
```

Comment:
Maybe we should use an assert here so that this has no overhead in a non debug build